### PR TITLE
specify heap size settings for modules; avm-server 0.12

### PR DESCRIPTION
**Pull request by justprosh**
- bump fluence-app-service to 0.11.1
- bump avm-server to 0.12.0
- add aquavm_max_heap_size, module_max_heap_size and module_default_heap_size to node's config
- check max_heap_size overflow in `add_module`

### CODE REVIEW

File: crates/server-config/src/node_config.rs [hunk 1]
```diff
@@ -82,6 +85,16 @@ pub struct NodeConfig {
     #[serde(default = "default_aquavm_pool_size")]
     pub aquavm_pool_size: usize,
 
+    /// Maximum heap size in bytes available for the module.
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub aquavm_max_heap_size: Option<bytesize::ByteSize>,
```
Comment (folex) [inline crates/server-config/src/node_config.rs:91]: why is it aquavm? shouldn't it be marine?

File: particle-modules/src/modules.rs [hunk 2]
```diff
@@ -131,11 +142,34 @@ impl ModuleRepository {
...
     }
 
+    // set default if max_heap_size and mem_pages_count are not specified
+    fn check_module_heap_size(&self, config: &mut TomlFaaSNamedModuleConfig) -> Result<()> {
+        let heap_size = match (config.config.max_heap_size, config.config.mem_pages_count) {
+            (Some(heap_size), _) => Some(heap_size),
+            (None, Some(pages_count)) => {
+                Some(ByteSize::b(marine_utils::wasm_pages_to_bytes(pages_count)))
+            }
+            (None, None) => self.default_heap_size.clone(),
```
Comment (folex) [inline particle-modules/src/modules.rs:152]: why clone? ByteSize isn't Copy?

File: particle-modules/src/modules.rs [hunk 3]
```diff
@@ -131,11 +142,34 @@ impl ModuleRepository {
         }
     }
 
+    // set default if max_heap_size and mem_pages_count are not specified
+    fn check_module_heap_size(&self, config: &mut TomlFaaSNamedModuleConfig) -> Result<()> {
```
Comment (mikevoronov) [inline particle-modules/src/modules.rs:146]: We could decouple this function from Marine into utils to avoid copying. But I think that it's better to do that in a separate PR.

Comment (justprosh) [reply particle-modules/src/modules.rs:146]: agree

File: crates/server-config/src/node_config.rs [hunk 4]
```diff
@@ -85,15 +85,20 @@ pub struct NodeConfig {
     #[serde(default = "default_aquavm_pool_size")]
     pub aquavm_pool_size: usize,
 
-    /// Maximum heap size in bytes available for the module.
+    /// Maximum heap size in bytes available for the VM.
```
Comment (folex) [inline crates/server-config/src/node_config.rs:88]:
```suggestion
    /// Maximum heap size in bytes available for an interpreter instance.
```

File: crates/server-config/src/node_config.rs [hunk 5]
```diff
@@ -85,15 +85,20 @@ pub struct NodeConfig {
...
     #[serde_as(as = "Option<DisplayFromStr>")]
     #[serde(default)]
     pub aquavm_max_heap_size: Option<bytesize::ByteSize>,
 
+    /// Maximum heap size in bytes available for the module.
```
Comment (folex) [inline crates/server-config/src/node_config.rs:93]:
```suggestion
    /// Maximum heap size in bytes available for a WASM module.
```

File: crates/server-config/src/node_config.rs [hunk 6]
```diff
@@ -85,15 +85,20 @@ pub struct NodeConfig {
...
+    #[serde_as(as = "Option<DisplayFromStr>")]
+    #[serde(default)]
+    pub app_service_max_heap_size: Option<bytesize::ByteSize>,
```
Comment (folex) [inline crates/server-config/src/node_config.rs:96]: Since not a service-wise setting, it's applied to each module individually. Maybe rename like this?
```suggestion
    pub service_module_max_heap: Option<bytesize::ByteSize>,
```

or `wasm_module_max_heap` or `module_max_heap`

File: crates/server-config/src/node_config.rs [hunk 7]
```diff
@@ -85,15 +85,20 @@ pub struct NodeConfig {
...
+
     /// Default heap size in bytes available for the module unless otherwise specified.
     #[serde_as(as = "Option<DisplayFromStr>")]
     #[serde(default)]
-    pub aquavm_default_heap_size: Option<bytesize::ByteSize>,
+    pub app_service_default_heap_size: Option<bytesize::ByteSize>,
```
Comment (folex) [inline crates/server-config/src/node_config.rs:101]: ditto

File: particle-modules/src/modules.rs [hunk 8]
```diff
@@ -149,7 +149,7 @@ impl ModuleRepository {
             (None, Some(pages_count)) => {
                 Some(ByteSize::b(marine_utils::wasm_pages_to_bytes(pages_count)))
             }
-            (None, None) => self.default_heap_size.clone(),
+            (None, None) => self.default_heap_size,
         };
 
         config.config.max_heap_size = heap_size.clone();
```
Comment (folex) [inline particle-modules/src/modules.rs:155]: why clone?

**Review by folex — APPROVED** [review]
